Selection List refactor phase 3: base

thiagobrez opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Base PR for the Phase 3 of the Selection List refactor.

Tracker issue: https://github.com/Expensify/App/issues/11795
Child issue: https://github.com/Expensify/App/issues/20354

- In `BaseSelectionList`:
  - Replace `ArrowKeyFocusManager` with `useArrowKeyFocusManager` (same behavior but hooks)
  - Add `footerContent`, `disableKeyboardShortcuts` and `children` props for lists that are gonna need it
- Create a new generic list item: `BaseListItem`. This component holds the styles to Checkbox (multiple selection) or Checkmark (single selection). Then, the existing `RadioListItem` and `UserListItem` represent just the contents of the list item, having a better separation of concerns
- In `UserListItem`, rename `item.avatar` to `item.icons`, so we can have multiple icon sources when the list item is a workspace
- Add optional prop `showTooltip` to `SubscriptAvatar`, to render the tooltip conditionally

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/20354
PROPOSAL: 


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

All lists using `SelectionList` should continue working as before. We can check a few of each type (single selection and multiple selection) to make sure both list items are functioning correctly:
- [ ] Profile -> Timezone
- [ ] Profile -> Pronouns
- [ ] Workspace -> Members
- [ ] Workspace -> Members -> Invite

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

All lists using `SelectionList` should continue working as before. We can check a few of each type (single selection and multiple selection) to make sure both list items are functioning correctly:
- [ ] Profile -> Timezone
- [ ] Profile -> Pronouns
- [ ] Workspace -> Members
- [ ] Workspace -> Members -> Invite

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/d4ff5ec6-1a85-470d-a4b1-e9dcd6e2c8d0" />

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/943a2c16-fd2a-4582-972c-9db423cb3143" />

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/4db9c805-1f84-4414-805b-f1df8f841815" />

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/437f929c-0272-4d7e-b788-f22a1a492f3c" />

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/c0c96cd6-be5f-48a5-83b1-b2b75f2915af" />

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

<video src="https://github.com/Expensify/App/assets/26878038/9f450356-5946-4ce7-8713-8c0bef14a7bf" />

</details>

thiagobrez committed:
chore: wip

thiagobrez committed:
Merge branch 'main' into thiagobrez/selection-list-phase-3-base

thiagobrez committed:
refactor(selection-list): create BaseListItem

thiagobrez committed:
chore: resolve conflicts

thiagobrez committed:
feat(selection-list): request money

thiagobrez committed:
chore: resolve conflicts

thiagobrez committed:
chore: rollback request money and split bill changes

thiagobrez committed:
chore: rollback request money and split bill changes

thiagobrez committed:
chore: rollback request money and split bill changes

thiagobrez committed:
chore: fix PropTYpes

thiagobrez committed:
chore: fix import

thiagobrez commented on src/components/SelectionList/BaseListItem.js:
```diff
@@ -0,0 +1,100 @@
+import OfflineWithFeedback from '../OfflineWithFeedback';
+import CONST from '../../CONST';
+
+function BaseListItem({item, isFocused = false, isDisabled = false, showTooltip, canSelectMultiple, onSelectRow, onDismissError = () => {}}) {
```
Previously the Checkbox (multiple selection) was located in `UserListItem`, and the Checkmark (single selection) in `RadioListItem`. Moved them both here, rendering conditionally. Then, the 2 mentioned list items can remain focused on rendering the content, with a better separation of concerns

rezkiy37 replied:
I see that `canSelectMultiple` is not required.
```suggestion
function BaseListItem({item, isFocused = false, isDisabled = false, showTooltip, canSelectMultiple = true/false, onSelectRow, onDismissError = () => {}}) {
```

thiagobrez commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -166,7 +165,20 @@ function BaseSelectionList({
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+
+    const [focusedIndex, setFocusedIndex] = useArrowKeyFocusManager({
+        maxIndex: flattenedSections.allOptions.length - 1,
+        onFocusedIndexChange: scrollToIndex,
+        initialFocusedIndex: _.findIndex(flattenedSections.allOptions, (option) => option.keyForList === initiallyFocusedOptionKey),
+        disabledIndexes: flattenedSections.disabledOptionsIndexes,
+        isActive: !disableKeyboardShortcuts,
+    });
```
Switched from `<ArrowKeyFocusManager />` to `useArrowKeyFocusManager`, no change of behavior

thiagobrez commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -290,121 +301,114 @@ function BaseSelectionList({
     });
 
     return (
-        <ArrowKeyFocusManager
```
This huge block of changes is actually just removing `<ArrowKeyFocusManager />` that was wrapping everything

thiagobrez commented on src/components/SelectionList/RadioListItem.js:
```diff
@@ -1,52 +1,18 @@
 import {radioListItemPropTypes} from './selectionListPropTypes';
 
-function RadioListItem({item, isFocused = false, isDisabled = false, onSelectRow}) {
+function RadioListItem({item, isFocused = false}) {
```
Removed the Checkmark part and moved to `BaseListItem`

thiagobrez commented on src/components/SelectionList/UserListItem.js:
```diff
@@ -47,62 +35,14 @@ function UserListItem({item, isFocused = false, showTooltip, onSelectRow, onDism
     );
 
     return (
-        <OfflineWithFeedback
```
Removed the Checkbox part and moved to `BaseListItem`

thiagobrez commented on src/components/SubscriptAvatar.js:
```diff
@@ -34,6 +37,7 @@ const defaultProps = {
+    showTooltip: true,
 };
 
 function SubscriptAvatar(props) {
```
Here I only put the components in variables so we can conditionally render the tooltip and avoid duplicating code

thiagobrez commented on src/components/SubscriptAvatar.js:
```diff
@@ -47,45 +51,60 @@ function SubscriptAvatar(props) {
         marginStyle = {};
     }
+
+    const mainAvatar = (
+        <Avatar
+            containerStyles={StyleUtils.getWidthAndHeightStyle(StyleUtils.getAvatarSize(props.size || CONST.AVATAR_SIZE.DEFAULT))}
+            source={props.mainAvatar.source}
+            size={props.size || CONST.AVATAR_SIZE.DEFAULT}
+            name={props.mainAvatar.name}
+            type={props.mainAvatar.type}
+        />
+    );
+
+    const secondaryAvatar = (
+        <View
+            style={[props.size === CONST.AVATAR_SIZE.SMALL_NORMAL ? styles.flex1 : {}, isSmall ? styles.secondAvatarSubscriptCompact : subscriptStyle]}
+            // Hover on overflowed part of icon will not work on Electron if dragArea is true
+            // https://stackoverflow.com/questions/56338939/hover-in-css-is-not-working-with-electron
+            dataSet={{dragArea: false}}
+        >
+            <Avatar
+                iconAdditionalStyles={[
+                    StyleUtils.getAvatarBorderWidth(isSmall ? CONST.AVATAR_SIZE.SMALL_SUBSCRIPT : CONST.AVATAR_SIZE.SUBSCRIPT),
+                    StyleUtils.getBorderColorStyle(props.backgroundColor),
+                ]}
+                source={props.secondaryAvatar.source}
+                size={isSmall ? CONST.AVATAR_SIZE.SMALL_SUBSCRIPT : CONST.AVATAR_SIZE.SUBSCRIPT}
+                fill={themeColors.iconSuccessFill}
+                name={props.secondaryAvatar.name}
+                type={props.secondaryAvatar.type}
+            />
+        </View>
+    );
```
Here I only put the components in variables so we can conditionally render the tooltip and avoid duplicating code

thiagobrez committed:
chore: rollback deleted function

rezkiy37 commented on src/components/SelectionList/BaseListItem.js:
```diff
@@ -0,0 +1,100 @@
+                onPress={() => onSelectRow(item)}
+                disabled={isDisabled}
+                accessibilityLabel={item.text}
+                accessibilityRole="button"
```
```suggestion
                accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
```

rezkiy37 commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -166,7 +165,20 @@ function BaseSelectionList({
+    const [focusedIndex, setFocusedIndex] = useArrowKeyFocusManager({
+        maxIndex: flattenedSections.allOptions.length - 1,
+        onFocusedIndexChange: scrollToIndex,
+        initialFocusedIndex: _.findIndex(flattenedSections.allOptions, (option) => option.keyForList === initiallyFocusedOptionKey),
```
Should we handle when `findIndex` returns `-1`?

thiagobrez replied:
Nope, it's intended. When `-1` the highlight will be disabled initially

rezkiy37 commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -290,121 +301,114 @@ function BaseSelectionList({
+                                    style={[styles.peopleRow, styles.userSelectNone, styles.ph5, styles.pb3]}
+                                    onPress={onSelectAll}
+                                    accessibilityLabel={translate('workspace.people.selectAll')}
+                                    accessibilityRole="button"
```
```suggestion
                                    accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
```

rezkiy37 commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -290,121 +301,114 @@ function BaseSelectionList({
+                                onScrollBeginDrag={onScrollBeginDrag}
+                                keyExtractor={(item) => item.keyForList}
+                                extraData={focusedIndex}
+                                indicatorStyle="white"
```
```suggestion
                                indicatorStyle={themeColors.white}
```

rezkiy37 commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -290,121 +301,114 @@ function BaseSelectionList({
+                                viewabilityConfig={{viewAreaCoveragePercentThreshold: 95}}
+                                testID="selection-list"
+                                style={[styles.flexGrow0]}
+                                onLayout={() => {
```
Can we extract it into a callback like a handle function?

thiagobrez replied:
Sure thing!

rezkiy37 reviewed (commented):
Left a few comments.

gedu commented on src/components/SubscriptAvatar.js:
```diff
@@ -47,45 +51,60 @@ function SubscriptAvatar(props) {
         marginStyle = {};
     }
+
+    const mainAvatar = (
```
Question, can be `MainAvatar` and below use it as
```
<MainAvatar />
```
I feel it odd seeing it like that 
same for `secondaryAvatar`
Also maybe creating them outside of the component (?)

thiagobrez replied:
Yeah, I also find it a little bit odd, but I'm just following the patterns that already exist.

Ideally I would create outside a wrapper component that changes the container conditionally, but I think it's better to stick with the existing patterns in this case

gedu commented on src/components/SubscriptAvatar.js:
```diff
@@ -34,6 +37,7 @@ const defaultProps = {
+    showTooltip: true,
 };
 
 function SubscriptAvatar(props) {
```
can we use destructuring props?

thiagobrez replied:
Sure!

thiagobrez committed:
chore: address pr comments

rezkiy37 reviewed (commented):
LGTM!

thiagobrez commented:
@ntdiary Hey! Yes, that sounds good :D

thiagobrez commented:
@ntdiary Hey! Don't worry, I'll be OOO from tomorrow until Tuesday, so it should give you (or whoever takes this) sufficient time. I'll resolve the conflict soon, but I guess you can already take a look when you have time because it's on `themes` so should be very simple to solve!

thiagobrez commented:
@ntdiary Actually, just talked with @ArekChr and he can take this as C+ on Monday

thiagobrez committed:
chore: fix conflicts

thiagobrez commented:
@ntdiary Thanks! Yeah, might be related to this one. I'll have to take a look at it next week though. Also, just sent the conflict fix!

thiagobrez committed:
chore: resolve conflicts

thiagobrez committed:
chore: rename variable

thiagobrez commented:
@ntdiary Update: @ArekChr is OOO for the week for medical reasons, so I asked @burczu to review if he has the bandwidth. He will see tomorrow, but if he can't I'll ping you again :)

Tomorrow I'll work on another higher-priority issue, but I'd appreciate if we could review the PR even with the small animation bug, so if more things to fix come up I can work on them all at once.

I'll keep you updated!

burczu commented:
@ntdiary Please assign me to this issue - I'll review it as C+. Thanks.

burczu commented:
@ntdiary Yeah, sure - I'm just wondering if my approval will trigger automation and assign internal engineer to review after me. We'll see ;)

burczu commented on src/components/SelectionList/selectionListPropTypes.js:
```diff
@@ -30,33 +52,25 @@ const userListItemPropTypes = {
         rightElement: PropTypes.element,
 
         /** Avatar for the user */
```
This comment was for `avatar` property - is it still applying for the new property `icons`?

thiagobrez replied:
Yep, still applies, but I guess I can make it more clear by matching the key name

burczu commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -259,33 +280,32 @@ function BaseSelectionList({
         );
     };
 
+    const onLayout = () => {
```
We should avoid naming event handlers like `onSomething` - it's even in PR Author's Checklist:

> I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. toggleReport and not onIconClick)

So the name should define what the handler does.

thiagobrez replied:
True, force of habit :)

Changing to `scrollToFocusedIndex`

burczu reviewed (commented):
@thiagobrez Added some comments.

burczu commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

burczu commented:
@thiagobrez I think I might find a bug on mobile web Safari - please take a look at the video below:

https://github.com/Expensify/App/assets/7771498/ada6730f-4f05-472f-bfc9-198ccbc6095d

thiagobrez commented on src/components/SubscriptAvatar.js:
```diff
@@ -34,60 +37,76 @@ const defaultProps = {
+                fallbackIcon={secondaryAvatar.fallbackIcon}
+            />
+        </View>
+    );
```
Yeah, I do not like it either, but did this way because I saw in other components and thought it was the project pattern. Ideally I would create a separate component on top of this one (not a new file), because this is being used just to conditional render the avatar with or without the tooltip wrapper. WDYT?

thiagobrez commented:
@burczu Interesting, yes I can confirm. I'll investigate. Thanks!

thiagobrez committed:
chore: remove useArrowKeyFocusManager hook in favor of component

thiagobrez commented:
@burczu Both the lack of animation bug and Timezone page showing empty list were being caused by switching from `<ArrowKeyFocusManager />` to the hook equivalent `useArrowKeyFocusManager`.

I rolled that back and it's fixed. Also sent the fixes to the other discussions!

thiagobrez committed:
chore: fix comment

thiagobrez committed:
refactor(tooltip): add ability to prevent render

thiagobrez commented:
@burczu Following on your first comment, I've removed the components in variables and refactored both `Tooltip` and `UserDetailsTooltip` to accept a prop `shouldRender` that behaves the same way.

Changes will look big, but in this is what was done:
- Rename `Tooltip/index.js` -> `Tooltip/BaseTooltip.js`
- Rename `Tooltip/index.native.js` -> `Tooltip/BaseTooltip.native.js`
- Create a new file `Tooltip/index.js` that renders the base component conditionally based on the `shouldRender` prop
---
- Rename `UserDetailsTooltip/index.js` -> `UserDetailsTooltip/BaseUserDetailsTooltip.js`
- Rename `UserDetailsTooltip/index.web.js` -> `UserDetailsTooltip/BaseUserDetailsTooltip.web.js`
- Create a new file `UserDetailsTooltip/index.js` that renders the base component conditionally based on the `shouldRender` prop

thiagobrez committed:
chore: fix tooltip props

thiagobrez committed:
test: fix formatMemberForList tests

burczu commented on src/components/UserDetailsTooltip/index.js:
```diff
@@ -1,19 +1,35 @@
-    /** Children to wrap with Tooltip. */
-    children: PropTypes.node.isRequired,
+    ...userDetailsTooltipPropTypes,
+    shouldRender: PropTypes.bool,
```
Missing comment description for `shouldRender`

thiagobrez replied:
Fixed!

burczu commented on src/components/Tooltip/index.js:
```diff
@@ -1,168 +1,35 @@
+
+const propTypes = {
+    ...tooltipPropTypes,
+    shouldRender: PropTypes.bool,
```
same here

thiagobrez replied:
Fixed!

thiagobrez committed:
chore: add proptypes comments

thiagobrez committed:
chore: add proptypes comments

burczu reviewed (approved):
LGTM

cristipaval reviewed (approved):
Thank you all for this work! 👏

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6383232884) to staging by https://github.com/cristipaval in version: 1.3.76-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6395179660) to production by https://github.com/mountiny in version: 1.3.76-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6395186072) to staging by https://github.com/cristipaval in version: 1.3.77-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6412851562) to production by https://github.com/mountiny in version: 1.3.77-7 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

robertKozik commented on src/pages/workspace/WorkspaceMembersPage.js:
```diff
@@ -322,11 +322,13 @@ function WorkspaceMembersPage(props) {
-                    name: details.login,
-                    type: CONST.ICON_TYPE_AVATAR,
-                },
+                icons: [
+                    {
+                        source: UserUtils.getAvatar(details.avatar, accountID),
+                        name: details.login,
+                        type: CONST.ICON_TYPE_AVATAR,
+                    },
+                ],
```
We forgot to pass the id of account as separate field in object https://github.com/Expensify/App/issues/29900#issuecomment-1769124767

rushatgabhane commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -169,7 +172,10 @@ function BaseSelectionList({
+
+        // If we don't disable dependencies here, we would need to make sure that the `sections` prop is stable in every usage of this component.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
```
This change caused a bug - [WS - Members - List of users invited to the workspace does not scroll with Up and Down arrows](https://github.com/Expensify/App/issues/29022#top)

`allOptions` is initially empty. But this callback runs only on mount, so it doesn't change when allOptions are populated.

burczu commented on src/components/SelectionList/BaseSelectionList.js:
```diff
@@ -259,33 +273,40 @@ function BaseSelectionList({
         );
     };
 
+    const scrollToFocusedIndexOnFirstRender = useCallback(() => {
```
It looks like introducing this caused issue: https://github.com/Expensify/App/issues/31660 (fixed by this PR: https://github.com/Expensify/App/pull/31737)

s77rt commented on src/components/SelectionList/RadioListItem.js:
```diff
@@ -1,51 +1,18 @@
-                        {item.text}
-                    </Text>
+        <View style={[styles.flex1, styles.alignItemsStart]}>
+            <Text style={[styles.optionDisplayName, isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText, item.isSelected && styles.sidebarLinkTextBold]}>{item.text}</Text>
```
We had to specify the `numberOfLines={1}` to show ellipsis if the text is too long instead of having it overflow
